### PR TITLE
lesson 5: Documentation

### PR DESCRIPTION
**Pull request by Duologic**
This adds a lesson on providing documentation.

As I was working through it also made a few minor cosmetic adjustments.

### CODE REVIEW

File: lessons/lesson5/lesson.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
+### Writing docstrings
+
+We'll continue with the webserver library from the exercise.
+
+%(example1/example1.jsonnet)s
+
+This library provides a number of functions to create a webserver. In docsonnet lingo this
```
Comment (julienduchesne) [inline lessons/lesson5/lesson.md:7]:
```suggestion
This library provides a number of functions to create a webserver. In docsonnet lingo, this
```

Comment (julienduchesne) [inline lessons/lesson5/lesson.md:7]: Add a link to docsonnet here?

File: lessons/lesson5/lesson.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+is called a 'package'. Each function has a few arguments which usually have a fixed type,
+as Jsonnet itself is not a typed language we'll simply try to document the type.
```
Comment (julienduchesne) [inline lessons/lesson5/lesson.md:9]: Punctuation is a bit weird here
```suggestion
is called a 'package'. Each function has a few arguments which usually have a fixed type;
As Jsonnet itself is not a typed language, we'll simply try to document the type.
```

File: lessons/lesson5/lesson.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+Note that the [docs for
+doc-util](https://github.com/jsonnet-libs/docsonnet/blob/master/doc-util/README.md) are
+also rendered by docsonnet.
+
+---
+
+%(example1/example2.jsonnet)s
+
+The docsonnet puts a claim on keys that start with a hash `#`. It assumes that keys
+starting with a hash symbol is very uncommon, additionally it closely relates to how
```
Comment (julienduchesne) [inline lessons/lesson5/lesson.md:27]:
```suggestion
starting with a hash symbol is very uncommon. Additionally, it closely relates to how
```

File: lessons/lesson5/lesson.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+The package definition is put in the `#` key without a value.
+
+The `url` argument is the part that can be used to install this library with
+jsonnet-bundler:
+
+`$ jb install <url>`
+
+In combination with the `url`, the `filename` refers to what should be imported, note the
+neat `std.thisFile` shortcut so we don't have to remember to change the filename here if
+we do so.
```
Comment (julienduchesne) [inline lessons/lesson5/lesson.md:39]: "if we do so". I don't really understand this sentence

File: lessons/lesson5/lesson.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+`webserver.withImage(webserver.images.nginx)`
+
+---
+
+%(example1/example5.jsonnet)s
+
+Just like with functions, the key for the docstring gets prefixed by `#`. Objects can be
+documented with `d.obj(help)`.
+
+The constants in this object can be documented with `d.val(type, help)`.
```
Comment (julienduchesne) [inline lessons/lesson5/lesson.md:74]: They're not constants. I'd call them "attributes", personally

File: lessons/lesson5/lesson.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+Or in short:
+
+`$ jsonnet -S -c -m docs/ example7.jsonnet`
+
+Note that this overwrites but does not remove existing files. 
+
+%(example1/Makefile)s
+
+A simple Makefile target can be quite useful to contain these incantations.
```
Comment (julienduchesne) [inline lessons/lesson5/lesson.md:112]: Nothing wrong here. Just had a laugh at "incantations" 😆

**Review by julienduchesne — APPROVED** [review]
👍

File: lessons/lesson5/lesson.md (hunk 1 above)
Comment (Duologic) [reply lessons/lesson5/lesson.md:7]: The intro for this page is in `main.jsonnet` which has a link.

File: lessons/lesson5/lesson.md (hunk 4 above)
Comment (Duologic) [reply lessons/lesson5/lesson.md:39]:
```suggestion
neat `std.thisFile` shortcut so we don't have to remember to change the filename here if
we ever rename the file.
```

**Pull request merged**